# Data Cleaning with Pandas

In this notebook we'll go through a few basic data cleaning steps that should be performed on all new datasets where necessary.

We'll go through the process with both the `orders` and `orderlines` datasets. You can then practice these skills by cleaning the `products` dataset yourself

In [1]:
import pandas as pd

In [2]:
# orders.csv
url = "https://drive.google.com/file/d/1Vu0q91qZw6lqhIqbjoXYvYAQTmVHh6uZ/view?usp=sharing"
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
orders = pd.read_csv(path)

# orderlines.csv
url = "https://drive.google.com/file/d/1FYhN_2AzTBFuWcfHaRuKcuCE6CWXsWtG/view?usp=sharing"
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
orderlines = pd.read_csv(path)

Before we begin, let's create a copy of the `orders` and `orderlines` DataFrames. This way we are sure any of our changes won't affect the original DataFrames

In [61]:
orders_df = orders.copy()

orderlines_df = orderlines.copy()

One of the best ways to begin data cleaning is by exploring using `.info()`. This will tell us:
* The shape of the DataFrame
* The names of the columns
* If there are any missing values
* The datatypes of the columns

By exploring the missing values and correcting any incorrect datatypes, we often come across inconsistencies in our data.

Beyond this, we should also have a **check for any duplicate rows**.

Let's first deal with the duplicates, as it's nice and easy, then we'll explore what `.info()` has to tell us.

## 1.&nbsp; Duplicates
We can check for duplicates using the pandas [.duplicated()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.duplicated.html) method.

We can then delete these rows, if we wish, using [.drop_duplicates()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop_duplicates.html)

In [137]:
# orders_df
orders_df.duplicated().sum()

0

In [63]:
# orderlines_df
orderlines_df.duplicated().sum()

0

We have no duplicate rows in either DataFrame. Easy, there is no problem to solve. Normally though, if there were some duplicates, we'd drop the extra rows.

In [64]:
products = pd.read_csv('products.csv')
products_df = products.copy()
products_df.info()
products_df.loc[products_df.duplicated()]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19326 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          19326 non-null  object
 1   name         19326 non-null  object
 2   desc         19319 non-null  object
 3   price        19280 non-null  object
 4   promo_price  19326 non-null  object
 5   in_stock     19326 non-null  int64 
 6   type         19276 non-null  object
dtypes: int64(1), object(6)
memory usage: 1.0+ MB


sku                                               name  \
101    APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
102    APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
103    APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
104    APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
105    APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
...        ...                                                ...   
16831  APP2302  Apple MacBook Pro 13 "Core i5 Touch Bar 33GHz ...   
16833  APP2303  Apple MacBook Pro 13 "Core i5 Touch Bar 33GHz ...   
18190  PAR0077                         Parrot Bebop Drone 2 Power   
18308  NKI0010              Nokia Wireless sphygmomanometer Plata   
18514  PLA0035          Bluetooth Headset Plantronics Explorer 80   

                                                    desc        price  \
101          MacBook Pro laptop 133 inches (MD101Y / A).         1199   
102          MacBook Pro laptop 133 inches (MD101Y / A).         1199   
103          MacBook Pro laptop 133 inches (MD101Y / A).         1199   
104          MacBook Pro laptop 133 inches (MD101Y / A).         1199   
105          MacBook Pro laptop 133 inches (MD101Y / A).         1199   
...                                                  ...          ...   
16831  New MacBook Pro 13-inch Core i5 Touch Bar 33 G...   26.155.941   
16833  New MacBook Pro 13 inch Touch Bar 33 GHz Core ...  237.559.421   
18190  Drone cuadricóptero quality camera integrated ...        699.9   
18308     Sphygmomanometer for iPhone iPad and iPod App.       129.99   
18514  egonómico wireless headset design for iPhone i...        39.99   

      promo_price  in_stock      type  
101    11.455.917         0      1282  
102    11.455.917         0      1282  
103    11.455.917         0      1282  
104    11.455.917         0      1282  
105    11.455.917         0      1282  
...           ...       ...       ...  
16831  26.155.941         0  1,02E+12  
16833  23.755.942         0  1,02E+12  
18190   6.733.892         0  11905404  
18308   1.149.899         1  11905404  
18514     349.896         1      5384  

[8746 rows x 7 columns]

# 2.&nbsp; `.info()`

In [386]:
orders_df.info()
orders_df.total_paid.isna().sum()

<class 'pandas.core.frame.DataFrame'>
Index: 226904 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      226904 non-null  int64         
 1   created_date  226904 non-null  datetime64[ns]
 2   total_paid    226904 non-null  float64       
 3   state         226904 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(1)
memory usage: 8.7+ MB


0

In [66]:
orders_df.order_id.nunique()

226909

* `total_paid` has 5 missing values
* `created_date` should become datetime datatype

In [67]:
orderlines_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB


* `date` should be a datetime datatype
* `unit_price` should be a float datatype

## 3.&nbsp; Missing values

### 3.1.&nbsp; Orders
* `total_paid` has 5 missing values

In [68]:
orders_df.isna().sum()

order_id        0
created_date    0
total_paid      5
state           0
dtype: int64

In [69]:
orders_df.loc[orders_df.total_paid.isna()]

order_id         created_date  total_paid    state
127701    427314  2017-11-20 18:54:39         NaN  Pending
132013    431655  2017-11-22 12:15:24         NaN  Pending
147316    447411  2017-11-27 10:32:37         NaN  Pending
148833    448966  2017-11-27 18:54:15         NaN  Pending
149434    449596  2017-11-27 21:52:08         NaN  Pending

In [70]:
print(f"5 missing values represents {((orders_df.total_paid.isna().sum() / orders_df.shape[0])*100).round(5)}% of the rows in our DataFrame")

5 missing values represents 0.0022% of the rows in our DataFrame


A quick way to find out a percentage here, if you don't need to print out a sentence for yourself/students/colleagues is `.value_count(normalize=True)`

In [71]:
orders_df.total_paid.isna().value_counts(normalize=True)

total_paid
False    0.999978
True     0.000022
Name: proportion, dtype: float64

As there is such a tiny amount of missing values, we will simply delete these rows, as we have enough data without them.

In [72]:
orders_df = orders_df.loc[~orders.total_paid.isna(), :]

In [73]:
# alternative method to get rid of null values:
orders_df.dropna()

order_id         created_date  total_paid            state
0         241319  2017-01-02 13:35:40       44.99        Cancelled
1         241423  2017-11-06 13:10:02      136.15        Completed
2         242832  2017-12-31 17:40:03       15.76        Completed
3         243330  2017-02-16 10:59:38       84.98        Completed
4         243784  2017-11-24 13:35:19      157.86        Cancelled
...          ...                  ...         ...              ...
226904    527397  2018-03-14 13:56:38       42.99      Place Order
226905    527398  2018-03-14 13:57:25       42.99  Shopping Basket
226906    527399  2018-03-14 13:57:34      141.58  Shopping Basket
226907    527400  2018-03-14 13:57:41       19.98  Shopping Basket
226908    527401  2018-03-14 13:58:36       18.98      Place Order

[226904 rows x 4 columns]

Should you have a significant number of missing values in the future, you have a choice:
+ you can impute the values
+ you can take the values from other DataFrames, if they are present there
+ you can delete the values
+ or any number of other creative solutions

Please, always consider how much time you have on your project, and what impact your method of choice will have on your final assesment.

In [74]:
orders_df.total_paid.isna().sum()

0

### 3.2.&nbsp; Orderlines
There are no missing values in `orderlines_df`

In [75]:
orderlines_df.isna().sum()

id                  0
id_order            0
product_id          0
product_quantity    0
sku                 0
unit_price          0
date                0
dtype: int64

## 4.&nbsp; Datatypes

In [77]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 226904 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      226904 non-null  int64  
 1   created_date  226904 non-null  object 
 2   total_paid    226904 non-null  float64
 3   state         226904 non-null  object 
dtypes: float64(1), int64(1), object(2)
memory usage: 8.7+ MB


### 4.1.&nbsp; Orders
* `created_date` should become datetime datatype

In [78]:
orders_df["created_date"] = pd.to_datetime(orders_df["created_date"])
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 226904 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      226904 non-null  int64         
 1   created_date  226904 non-null  datetime64[ns]
 2   total_paid    226904 non-null  float64       
 3   state         226904 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(1)
memory usage: 8.7+ MB


### 4.1.&nbsp; Orderlines
* `date` should be a datetime datatype
* `unit_price` should be a float datatype

In [79]:
orderlines_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB


#### 4.1.1.&nbsp; `date`

In [80]:
orderlines_df["date"] = pd.to_datetime(orderlines_df["date"])
orderlines_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                293983 non-null  int64         
 1   id_order          293983 non-null  int64         
 2   product_id        293983 non-null  int64         
 3   product_quantity  293983 non-null  int64         
 4   sku               293983 non-null  object        
 5   unit_price        293983 non-null  object        
 6   date              293983 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int64(4), object(2)
memory usage: 15.7+ MB


#### 4.1.2.&nbsp;`unit_price`

In [81]:
orderlines_df["unit_price"] = pd.to_numeric(orderlines_df["unit_price"])

ValueError: Unable to parse string "1.137.99" at position 6

As you can see when we try to convert `unit_price` to a numerical datatype, we receive a `ValueError` telling us that pandas doesn't understand the number `1.137.99`. This is probably because numbers cannot have 2 decimal points. Let's see if there are any other numbers like this.

In [86]:
orderlines_df.head(10)

id  id_order  product_id  product_quantity      sku unit_price  \
0  1119109    299539           0                 1  OTT0133      18.99   
1  1119110    299540           0                 1  LGE0043     399.00   
2  1119111    299541           0                 1  PAR0071     474.05   
3  1119112    299542           0                 1  WDT0315      68.39   
4  1119113    299543           0                 1  JBL0104      23.74   
5  1119114    295310           0                10  WDT0249     231.79   
6  1119115    299544           0                 1  APP1582   1.137.99   
7  1119116    299545           0                 1  OWC0100      47.49   
8  1119119    299546           0                 1  IOT0014      18.99   
9  1119120    295347           0                 1  APP0700      72.19   

                 date  
0 2017-01-01 00:07:19  
1 2017-01-01 00:19:45  
2 2017-01-01 00:20:57  
3 2017-01-01 00:51:40  
4 2017-01-01 01:06:38  
5 2017-01-01 01:14:27  
6 2017-01-01 01:17:21  
7 2017-01-01 01:46:16  
8 2017-01-01 01:50:34  
9 2017-01-01 01:54:11

In [88]:
orders_df[orders_df['order_id']==299544]

order_id        created_date  total_paid            state
206    299544 2017-01-01 01:17:21     1137.99  Shopping Basket

In [82]:
orderlines_df.unit_price.str.contains("\d+\.\d+\.\d+").value_counts()

unit_price
False    257814
True      36169
Name: count, dtype: int64

Looks like over 36000 rows in `orderlines` are affected by this problem. Let's work out how much that is as a percentage of our total data.

In [85]:
two_dot_percentage = ((orderlines_df.unit_price.str.contains("\d+\.\d+\.\d+").value_counts()[1] / orderlines_df.shape[0])*100).round(2)
print(f"The 2 dot problem represents {two_dot_percentage}% of the rows in our DataFrame")

The 2 dot problem represents 12.3% of the rows in our DataFrame


/tmp/ipykernel_9112/3688152030.py:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  two_dot_percentage = ((orderlines_df.unit_price.str.contains("\d+\.\d+\.\d+").value_counts()[1] / orderlines_df.shape[0])*100).round(2)


This is a bit of a tricky decision as 12.3% is a significant amount of our data... and we might even end up losing a larger portion of our data than this too. For the moment we will delete the rows as we only have 2 weeks for this project and I'd like some quick, accurate results to show. If we have time at the end, we can come back and investigate this problem further, maybe there's a solution?

Each row of `orderlines` represents a product in an order. For example, if order number 175 contained 3 seperate products, then order 175 would have 3 rows in `orderlines`, one row for each of the products. If 2 of those products have 'normal' prices (14.99, 15.85) and 1 has a price with 2 decimal points (1.137.99), we need to remove the whole order and not just the affected row. If we only remove the row with 2 decimal places then any later analysis about products and prices could be misleading.

We therefore need to find the order numbers associated with the rows that have 2 decimal points, and then remove all the associated rows.

In [175]:
two_dot_order_ids_list = orderlines_df.loc[orderlines_df.unit_price.str.contains("\d+\.\d+\.\d+"), "id_order"]#.unique()
print(type(two_dot_order_ids_list))
print(len(two_dot_order_ids_list))

<class 'pandas.core.series.Series'>
36169


In [173]:
orders_df[orders_df['order_id'].isin(two_dot_order_ids_list)]

order_id        created_date  total_paid            state
31        257847 2017-11-23 23:46:54     1367.11        Completed
35        258985 2017-07-31 12:52:38     2264.60        Completed
40        259668 2017-10-06 22:06:58     1132.33        Completed
46        262016 2017-08-18 01:05:38     3109.57        Completed
64        268372 2017-02-28 16:37:58     1243.53          Pending
...          ...                 ...         ...              ...
226793    527286 2018-03-14 12:55:14     1165.99      Place Order
226828    527321 2018-03-14 13:16:30     3503.99          Pending
226831    527324 2018-03-14 13:10:56     3075.00  Shopping Basket
226849    527342 2018-03-14 13:24:51     1335.99      Place Order
226871    527364 2018-03-14 13:36:13     1329.00  Shopping Basket

[34622 rows x 4 columns]

In [107]:
orderlines_df = orderlines_df.loc[~orderlines_df.id_order.isin(two_dot_order_ids_list)]

pandas.core.frame.DataFrame

In [475]:
orderlines_df.shape[0]

216250

In [618]:
293983 - orderlines_df.shape[0]

77733

In [477]:
orderlines_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 216250 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                216250 non-null  int64         
 1   id_order          216250 non-null  int64         
 2   product_id        216250 non-null  int64         
 3   product_quantity  216250 non-null  int64         
 4   sku               216250 non-null  object        
 5   unit_price        216250 non-null  float64       
 6   date              216250 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int64(4), object(1)
memory usage: 13.2+ MB


We still have 216250 rows in orderlines to work with. This should be more than enough for our evaluation.

Now that all of the 2 decimal point prices have been removed, let's try again to convert the column `unit_price` to the correct datatype.

In [616]:
orderlines_df["unit_price"] = pd.to_numeric(orderlines_df["unit_price"])

It worked perfectly

In [619]:
orderlines_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 216250 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                216250 non-null  int64         
 1   id_order          216250 non-null  int64         
 2   product_id        216250 non-null  int64         
 3   product_quantity  216250 non-null  int64         
 4   sku               216250 non-null  object        
 5   unit_price        216250 non-null  float64       
 6   date              216250 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int64(4), object(1)
memory usage: 13.2+ MB


# Challenge: Clean the `products` DataFrame
Now it's your turn. Use the lessons you learnt above and clean the products DataFrame. You don't have to copy exactly what we did. Think about the consequences of your actions, sometimes it is ok to delete rows, other times you may wish to come up with more creative solutions.

In [551]:
# products.csv
# url = "https://drive.google.com/file/d/1afxwDXfl-7cQ_qLwyDitfcCx3u7WMvkU/view?usp=sharing"
# path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
products = pd.read_csv('products.csv')

In [552]:
products.info()
products.sku.nunique()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19326 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          19326 non-null  object
 1   name         19326 non-null  object
 2   desc         19319 non-null  object
 3   price        19280 non-null  object
 4   promo_price  19326 non-null  object
 5   in_stock     19326 non-null  int64 
 6   type         19276 non-null  object
dtypes: int64(1), object(6)
memory usage: 1.0+ MB


10579

In [553]:
# your code here
prod_df = products.copy()
prod_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19326 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          19326 non-null  object
 1   name         19326 non-null  object
 2   desc         19319 non-null  object
 3   price        19280 non-null  object
 4   promo_price  19326 non-null  object
 5   in_stock     19326 non-null  int64 
 6   type         19276 non-null  object
dtypes: int64(1), object(6)
memory usage: 1.0+ MB


### Look for Duplicates

In [554]:
# your code here
prod_df.duplicated().sum()

8746

In [555]:
prod_df.duplicated().count() - prod_df.duplicated().sum()

10580

In [556]:
prod_df[prod_df.duplicated()].tail()

sku                                               name  \
16831  APP2302  Apple MacBook Pro 13 "Core i5 Touch Bar 33GHz ...   
16833  APP2303  Apple MacBook Pro 13 "Core i5 Touch Bar 33GHz ...   
18190  PAR0077                         Parrot Bebop Drone 2 Power   
18308  NKI0010              Nokia Wireless sphygmomanometer Plata   
18514  PLA0035          Bluetooth Headset Plantronics Explorer 80   

                                                    desc        price  \
16831  New MacBook Pro 13-inch Core i5 Touch Bar 33 G...   26.155.941   
16833  New MacBook Pro 13 inch Touch Bar 33 GHz Core ...  237.559.421   
18190  Drone cuadricóptero quality camera integrated ...        699.9   
18308     Sphygmomanometer for iPhone iPad and iPod App.       129.99   
18514  egonómico wireless headset design for iPhone i...        39.99   

      promo_price  in_stock      type  
16831  26.155.941         0  1,02E+12  
16833  23.755.942         0  1,02E+12  
18190   6.733.892         0  11905404  
18308   1.149.899         1  11905404  
18514     349.896         1      5384

In [557]:
prod_df[prod_df['sku']=='NKI0010']

sku                                   name  \
18307  NKI0010  Nokia Wireless sphygmomanometer Plata   
18308  NKI0010  Nokia Wireless sphygmomanometer Plata   

                                                 desc   price promo_price  \
18307  Sphygmomanometer for iPhone iPad and iPod App.  129.99   1.149.899   
18308  Sphygmomanometer for iPhone iPad and iPod App.  129.99   1.149.899   

       in_stock      type  
18307         1  11905404  
18308         1  11905404

In [558]:
prod_df = prod_df.drop_duplicates()
prod_df.count()

sku            10580
name           10580
desc           10573
price          10534
promo_price    10580
in_stock       10580
type           10530
dtype: int64

In [559]:
prod_df[prod_df['sku']=='NKI0010']

sku                                   name  \
18307  NKI0010  Nokia Wireless sphygmomanometer Plata   

                                                 desc   price promo_price  \
18307  Sphygmomanometer for iPhone iPad and iPod App.  129.99   1.149.899   

       in_stock      type  
18307         1  11905404

In [560]:
prod_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10580 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          10580 non-null  object
 1   name         10580 non-null  object
 2   desc         10573 non-null  object
 3   price        10534 non-null  object
 4   promo_price  10580 non-null  object
 5   in_stock     10580 non-null  int64 
 6   type         10530 non-null  object
dtypes: int64(1), object(6)
memory usage: 661.2+ KB


### Look for Missing values


In [577]:
# your code here
prod_df.type.isna().sum()
#prod_df[prod_df.type.isna()]

0

In [578]:
prod_df.type.isna().value_counts(normalize=True)

type
False    1.0
Name: proportion, dtype: float64

In [579]:
prod_df = prod_df.loc[~prod_df.type.isna()]

In [580]:
prod_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10530 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          10530 non-null  object
 1   name         10530 non-null  object
 2   desc         10530 non-null  object
 3   price        10484 non-null  object
 4   promo_price  10530 non-null  object
 5   in_stock     10530 non-null  int64 
 6   type         10530 non-null  object
dtypes: int64(1), object(6)
memory usage: 658.1+ KB


## description: we can either drop the 7 missing values or copying the name into description

In [566]:
prod_df.desc.isna().sum()

7

In [567]:
prod_df.desc.isna().value_counts(normalize=True)

desc
False    0.999335
True     0.000665
Name: proportion, dtype: float64

In [568]:
prod_df[prod_df.desc.isna()]

sku                                               name desc  \
16126  WDT0211-A  Open - Purple 2TB WD 35 "PC Security Mac hard ...  NaN   
16128  APP1622-A  Open - Apple Smart Keyboard Pro Keyboard Folio...  NaN   
17843    PAC2334             Synology DS718 + NAS Server | 10GB RAM  NaN   
18152  KAN0034-A  Open - Kanex USB-C Gigabit Ethernet Adapter Ma...  NaN   
18490    HTE0025  Hyper Pearl 1600mAh battery Mini USB Mirror an...  NaN   
18612    OTT0200    OtterBox External Battery Power Pack 20000 mAHr  NaN   
18690  HOW0001-A  Open - Honeywell thermostat Lyric zonificador ...  NaN   

           price promo_price  in_stock      type  
16126        107     814.659         0      1298  
16128  1.568.206   1.568.206         0      1298  
17843     566.35   5.659.896         0  12175397  
18152      29.99     237.925         0      1298  
18490      24.99       22.99         1      1515  
18612      79.99       56.99         1      1515  
18690     199.99   1.441.174         0  11905404

In [572]:
prod_df.loc[prod_df['sku']=='WDT0211-A']

sku                                               name desc  \
16126  WDT0211-A  Open - Purple 2TB WD 35 "PC Security Mac hard ...  NaN   

      price promo_price  in_stock  type  
16126   107     814.659         0  1298

In [574]:
prod_df.loc[prod_df.desc.isna(), 'desc'] = prod_df.loc[prod_df.desc.isna(), 'name']
prod_df.loc[prod_df['sku']=='WDT0211-A']

sku                                               name  \
16126  WDT0211-A  Open - Purple 2TB WD 35 "PC Security Mac hard ...   

                                                    desc price promo_price  \
16126  Open - Purple 2TB WD 35 "PC Security Mac hard ...   107     814.659   

       in_stock  type  
16126         0  1298

In [575]:
# drop the 7 values
#prod_df = prod_df.loc[~prod_df.desc.isna()]

In [576]:
prod_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10530 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          10530 non-null  object
 1   name         10530 non-null  object
 2   desc         10530 non-null  object
 3   price        10484 non-null  object
 4   promo_price  10530 non-null  object
 5   in_stock     10530 non-null  int64 
 6   type         10530 non-null  object
dtypes: int64(1), object(6)
memory usage: 658.1+ KB


In [581]:
prod_df.price.isna().sum()
#prod_df[prod_df.price.isna()]

46

In [582]:
prod_df.price.isna().value_counts(normalize=True)

price
False    0.995632
True     0.004368
Name: proportion, dtype: float64

In [583]:
prod_df = prod_df.loc[~prod_df.type.isna()]

In [584]:
prod_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10530 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          10530 non-null  object
 1   name         10530 non-null  object
 2   desc         10530 non-null  object
 3   price        10484 non-null  object
 4   promo_price  10530 non-null  object
 5   in_stock     10530 non-null  int64 
 6   type         10530 non-null  object
dtypes: int64(1), object(6)
memory usage: 658.1+ KB


### Check / Change Data types

In [585]:
# your code here
prod_df.price

0        59.99
1           59
2           59
3           25
4        34.99
         ...  
19321    29.99
19322    69.95
19323    69.95
19324    69.95
19325    69.95
Name: price, Length: 10530, dtype: object

In [615]:
prod_df.price = pd.to_numeric(prod_df.price)

ValueError: Unable to parse string "1.639.792" at position 503

In [590]:
prod_df.info()
prod_df.price.str.contains("\d+\.\d+\.\d+").value_counts()

<class 'pandas.core.frame.DataFrame'>
Index: 10530 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          10530 non-null  object
 1   name         10530 non-null  object
 2   desc         10530 non-null  object
 3   price        10484 non-null  object
 4   promo_price  10530 non-null  object
 5   in_stock     10530 non-null  int64 
 6   type         10530 non-null  object
dtypes: int64(1), object(6)
memory usage: 658.1+ KB


price
False    10110
True       374
Name: count, dtype: int64

In [588]:
prod_df.price.str.contains("\d+\.\d+\.\d+").value_counts(normalize=True)

price
False    0.964327
True     0.035673
Name: proportion, dtype: float64

In [609]:
prod_df.loc[prod_df.price.astype(str).str.contains("\d+\.\d{3,}")].head()

sku                                               name  \
362  REP0043                      Speaker lower repair iPhone 4   
480  PIE0011                     Internal Battery for iPhone 3G   
515  SEN0061  Sennheiser EZX 80 Handsfree iPhone iPad and iP...   
518  SEV0026  Service installation RAM + HDD + SSD MacBook /...   
525  SEV0024      Service installation RAM + HDD + SSD Mac mini   

                                                  desc    price promo_price  \
362  Repair service including parts and labor for i...  499.004     499.004   
480        Replacement AC Adapter for Apple iPhone 3G.   98.978      98.978   
515  IPhone bluetooth headset with microphone iPad ...  649.891     649.891   
518  RAM + HDD installation + SSD in your MacBook /...  599.918     599.918   
525  installation RAM HDD + SSD + on your Mac mini ...  599.918     599.918   

     in_stock      type  
362         0  1,44E+11  
480         0  21485407  
515         0      5384  
518         0  20642062  
525         0  20642062

In [610]:
prod_df.price.astype(str).str.contains("\d+\.\d{3,}").value_counts()

price
False    9992
True      538
Name: count, dtype: int64

In [612]:
prod_df.price.astype(str).str.contains("\d+\.\d{3,}").value_counts(normalize=True)

price
False    0.948908
True     0.051092
Name: proportion, dtype: float64

In [613]:
prod_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10530 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          10530 non-null  object
 1   name         10530 non-null  object
 2   desc         10530 non-null  object
 3   price        10484 non-null  object
 4   promo_price  10530 non-null  object
 5   in_stock     10530 non-null  int64 
 6   type         10530 non-null  object
dtypes: int64(1), object(6)
memory usage: 658.1+ KB


In [614]:
# we eliminate all the 374 rows with corrupted prices:
prod_df = prod_df.loc[~(prod_df['price'].astype(str).str.contains("\d+\.\d+\.\d+")) | (prod_df['price'].astype(str).str.contains("\d+\.\d{3,}")) ]
prod_df.info()

# Alternative: make a list of the relative sku with corrupted prices:
#two_dot_prod_df_list = prod_df.loc[prod_df.price.str.contains("\d+\.\d+\.\d+"), "sku"]#.unique()
#prod_df[prod_df['sku'].isin(two_dot_prod_df_list)]
#prod_df = prod_df.loc[~prod_df.sku.isin(two_dot_prod_df_list)]

<class 'pandas.core.frame.DataFrame'>
Index: 10530 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          10530 non-null  object
 1   name         10530 non-null  object
 2   desc         10530 non-null  object
 3   price        10484 non-null  object
 4   promo_price  10530 non-null  object
 5   in_stock     10530 non-null  int64 
 6   type         10530 non-null  object
dtypes: int64(1), object(6)
memory usage: 658.1+ KB


In [604]:
prod_df.price = pd.to_numeric(prod_df.price)

ValueError: Unable to parse string "1.639.792" at position 503

In [455]:
prod_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10104 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   sku          10104 non-null  object 
 1   name         10104 non-null  object 
 2   desc         10104 non-null  object 
 3   price        10104 non-null  float64
 4   promo_price  10104 non-null  object 
 5   in_stock     10104 non-null  int64  
 6   type         10104 non-null  object 
dtypes: float64(1), int64(1), object(5)
memory usage: 631.5+ KB


In [488]:
prod_df.type.value_counts().head(20)

type
11865403    1061
12175397     940
1298         818
11935397     565
11905404     463
1282         373
12635403     367
13835403     269
5,74E+15     247
1364         218
12585395     190
1296         188
1325         184
5384         180
1433         179
12215397     173
5398         160
1,44E+11     159
57445397     135
1,02E+12     130
Name: count, dtype: int64

In [490]:
prod_df.type.nunique()

126

In [472]:
prod_df.sku.nunique()

10104

In [459]:
prod_df.promo_price = pd.to_numeric(prod_df.promo_price)

ValueError: Unable to parse string "1.119.976" at position 49

In [461]:
prod_df.promo_price.str.contains("\d+\.\d+\.\d+").sum()

4284

In [464]:
prod_df.promo_price.str.contains("\d+\.\d+\.\d+").value_counts()

promo_price
False    5820
True     4284
Name: count, dtype: int64

In [465]:
prod_df.promo_price.str.contains("\d+\.\d+\.\d+").value_counts(normalize=True)

promo_price
False    0.57601
True     0.42399
Name: proportion, dtype: float64

In [474]:
prod_df[prod_df.promo_price.str.contains("\d+\.\d+\.\d+")].head(10)

sku                                               name  \
50   APP0367  Apple Mini DisplayPort to DVI Adapter Mac dual...   
51   APP0344          Apple Thunderbolt Display 27 "Monitor Mac   
66   MAK0008  Maclocks theft case iPad 2 3 and 4 black with ...   
67   MAK0007  Maclocks theft case iPad 2 3 and 4 transparent...   
97   MAK0014  Maclocks safety housing Kiosk iPad 2 3 and 4 b...   
100  APP0390  Apple MacBook Pro 133 "Core i5 25GHz | 4GB RAM...   
109  PAC0508  Apple MacBook Pro 133 '' 25GHz | 16GB RAM | 1T...   
118  PAC0507  Apple MacBook Pro 133 '' 25Ghz | 16GB RAM | Fu...   
127  PAC0515  Apple MacBook Pro 133 "i7 29GHz | RAM 16GB | 5...   
136  PAC0510  Apple MacBook Pro 133 "i7 29GHz | RAM 16GB | 7...   

                                                  desc    price promo_price  \
50   Adapter Mini Display Port to DVI dual channel ...   119.00   1.119.976   
51   Monitor Display 27-inch Apple Thunderbolt (MC9...  1149.00  10.449.923   
66   Case antitheft iPad 2 3 and 4 polycarbonate ro...   120.00   1.199.957   
67   Case antitheft iPad 2 3 and 4 polycarbonate ro...   120.00   1.079.961   
97   Holder and housing iPad 2 3 and 4 aluminum and...   164.99   1.649.896   
100        MacBook Pro laptop 133 inches (MD101Y / A).  1199.00  11.455.917   
109  Apple MacBook Pro Fusion Drive 16GB 2 internal...  1919.00  16.999.895   
118  Apple MacBook Pro Fusion Drive 16GB 2 internal...  1639.00  15.989.896   
127  Apple MacBook Pro 133 inches (MD101Y / A) and ...  2039.00  20.379.897   
136  Apple MacBook Pro Fusion Drive 16GB 2 internal...  2039.00  20.379.897   

     in_stock      type  
50          0      1325  
51          0      1296  
66          0  12635403  
67          0  12635403  
97          0      1216  
100         0      1282  
109         0      1282  
118         0      1282  
127         0      1282  
136         0      1282

In [491]:
# orderlines_df is the official clean df to work with:
orderlines_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 216250 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                216250 non-null  int64         
 1   id_order          216250 non-null  int64         
 2   product_id        216250 non-null  int64         
 3   product_quantity  216250 non-null  int64         
 4   sku               216250 non-null  object        
 5   unit_price        216250 non-null  float64       
 6   date              216250 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int64(4), object(1)
memory usage: 13.2+ MB


In [494]:
orderlines_df[orderlines_df.sku.duplicated()].sort_values(by='sku').head(30)

id  id_order  product_id  product_quantity        sku  \
124405  1366184    406717           0                 1  8MO0001-A   
287176  1640610    523426           0                 1  8MO0003-A   
281286  1631017    519204           0                 1  8MO0003-A   
110001  1328041    394230           0                 1    8MO0007   
87886   1287706    375111           0                 1    8MO0007   
61887   1242009    352909           0                 1    8MO0007   
89487   1290653    376383           0                 1    8MO0007   
110980  1329875    395135           0                 2    8MO0007   
88732   1289252    375729           0                 1    8MO0007   
111137  1330253    395265           0                 1    8MO0007   
82968   1278978    370939           0                 1    8MO0007   
103924  1317024    388985           0                 1    8MO0007   
58114   1235486    349725           0                 1    8MO0007   
102937  1315269    388140           0                 1    8MO0007   
102824  1315079    388037           0                 1    8MO0007   
69759   1255947    359574           0                 1    8MO0007   
100820  1311564    386308           0                 1    8MO0007   
82830   1278735    370833           0                 1    8MO0007   
112932  1345308    396610           0                 1    8MO0007   
82755   1278606    370756           0                 1    8MO0007   
83991   1280688    371783           0                 1    8MO0007   
82747   1278595    370754           0                 1    8MO0007   
82656   1278443    370682           0                 1    8MO0007   
55673   1231276    347721           0                 1    8MO0007   
100807  1311545    386296           0                 1    8MO0007   
112661  1344848    396397           0                 1    8MO0007   
56282   1232346    347814           0                 1    8MO0007   
72575   1260971    362001           0                 1    8MO0007   
56333   1232464    348319           0                 1    8MO0007   
56374   1232530    348356           0                 1    8MO0007   

        unit_price                date  
124405       12.96 2017-09-30 17:11:48  
287176       12.85 2018-03-08 18:52:29  
281286       12.85 2018-02-28 23:43:59  
110001       19.99 2017-08-30 07:47:14  
87886        19.99 2017-07-10 07:07:22  
61887        23.99 2017-05-05 10:03:54  
89487        19.99 2017-07-12 13:20:19  
110980       19.99 2017-08-31 17:58:22  
88732        19.99 2017-07-11 08:46:36  
111137       19.99 2017-09-01 01:28:57  
82968        19.99 2017-06-30 18:02:43  
103924       19.99 2017-08-12 16:14:19  
58114        23.99 2017-04-25 19:35:05  
102937       19.99 2017-08-10 11:18:13  
102824       19.99 2017-08-09 22:59:40  
69759        23.99 2017-05-27 09:09:08  
100820       19.99 2017-08-04 22:37:31  
82830        19.99 2017-06-30 14:16:17  
112932       19.99 2017-09-05 07:40:35  
82755        19.99 2017-06-30 12:59:54  
83991        19.99 2017-07-03 11:59:27  
82747        19.99 2017-06-30 12:53:57  
82656        19.99 2017-06-30 11:57:17  
55673        23.99 2017-04-20 10:20:02  
100807       19.99 2017-08-04 21:07:40  
112661       19.99 2017-09-04 15:35:23  
56282        23.99 2017-04-21 11:35:01  
72575        23.99 2017-06-03 18:06:45  
56333        23.99 2017-04-21 12:32:32  
56374        23.99 2017-04-21 13:03:45

In [484]:
prod_df[prod_df['sku']=='8MO0007']

sku                                               name  \
15216  8MO0007  8Mobility iSlice Micro SD adapter for Macbook ...   

                                                desc  price promo_price  \
15216  Micro SD card adapter for MacBook Air 13-inch   35.0     239.895   

       in_stock      type  
15216         0  12585395

In [500]:
import pandas as pd
import numpy as np

# Create a sample DataFrame
data = {'Column1': ['A', np.nan, 'C', np.nan, 'E'],
        'Column2': ['AAA', 'BBB', 'CCC', 'DDD', 'EEE']}
df = pd.DataFrame(data)

# Combine the values from Column1 and Column2 while replacing NaN values
df['Combined'] = df['Column1'].fillna(df['Column2'])

# Print the modified DataFrame
df.info()
df.Column1.isna()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Column1   3 non-null      object
 1   Column2   5 non-null      object
 2   Combined  5 non-null      object
dtypes: object(3)
memory usage: 248.0+ bytes


0    False
1     True
2    False
3     True
4    False
Name: Column1, dtype: bool

In [360]:
ord_l_merged.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 9 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   id                293983 non-null  int64  
 1   id_order          293983 non-null  int64  
 2   product_id        293983 non-null  int64  
 3   product_quantity  293983 non-null  int64  
 4   sku               293983 non-null  object 
 5   unit_price        293983 non-null  object 
 6   date              293983 non-null  object 
 7   order_id          73189 non-null   float64
 8   total_paid        293983 non-null  object 
dtypes: float64(1), int64(4), object(4)
memory usage: 20.2+ MB


In [505]:
ord_l_merged.loc[~ord_l_merged.total_paid.isna(), :].head(10)

id  id_order_x  product_id  product_quantity      sku unit_price  \
6    1119115      299544           0                 1  APP1582   1.137.99   
11   1119126      299549           0                 1  PAC0929   2.565.99   
15   1119131      299553           0                 1  APP1854   3.278.99   
43   1119195      299582           0                 1  PAC0961   2.616.99   
59   1119214      299596           0                 1  PAC1599   2.873.99   
61   1119216      299596           0                 1  PAC1044   2.780.99   
64   1119219      299596           0                 1  PAC1583   2.316.99   
105  1119294      299629           0                 1  QNA0175   2.054.99   
107  1119296      298391           0                 1  PAC1594   2.769.99   
109  1119299      298391           0                 1  APP0662     149.99   

                    date  order_id  total_paid  id_order_y  
6    2017-01-01 01:17:21  299544.0     1137.99    299544.0  
11   2017-01-01 02:07:42  299549.0     2565.99    299549.0  
15   2017-01-01 02:14:47  299553.0     3278.99    299553.0  
43   2017-01-01 08:54:00  299582.0     2616.99    299582.0  
59   2017-01-01 09:53:11  299596.0     8039.96    299596.0  
61   2017-01-01 09:55:45  299596.0     8039.96    299596.0  
64   2017-01-01 10:00:17  299596.0     8039.96    299596.0  
105  2017-01-01 11:39:13  299629.0     2054.99    299629.0  
107  2017-01-01 11:41:04  298391.0     2919.98    298391.0  
109  2017-01-01 11:41:58  298391.0     2919.98    298391.0

In [507]:
del ord_l_merged['id_order_y']

In [512]:
ord_l_merged['Combined'] = ord_l_merged['total_paid'].fillna(ord_l_merged['unit_price'])
ord_l_merged.head(10)

id  id_order_x  product_id  product_quantity      sku unit_price  \
0  1119109      299539           0                 1  OTT0133      18.99   
1  1119110      299540           0                 1  LGE0043     399.00   
2  1119111      299541           0                 1  PAR0071     474.05   
3  1119112      299542           0                 1  WDT0315      68.39   
4  1119113      299543           0                 1  JBL0104      23.74   
5  1119114      295310           0                10  WDT0249     231.79   
6  1119115      299544           0                 1  APP1582   1.137.99   
7  1119116      299545           0                 1  OWC0100      47.49   
8  1119119      299546           0                 1  IOT0014      18.99   
9  1119120      295347           0                 1  APP0700      72.19   

                  date  order_id  total_paid Combined  
0  2017-01-01 00:07:19       NaN         NaN    18.99  
1  2017-01-01 00:19:45       NaN         NaN   399.00  
2  2017-01-01 00:20:57       NaN         NaN   474.05  
3  2017-01-01 00:51:40       NaN         NaN    68.39  
4  2017-01-01 01:06:38       NaN         NaN    23.74  
5  2017-01-01 01:14:27       NaN         NaN   231.79  
6  2017-01-01 01:17:21  299544.0     1137.99  1137.99  
7  2017-01-01 01:46:16       NaN         NaN    47.49  
8  2017-01-01 01:50:34       NaN         NaN    18.99  
9  2017-01-01 01:54:11       NaN         NaN    72.19